flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java

```diff
@@ -101,84 +132,116 @@ public void teardown() throws Exception {
 		return env;
 	}
 
-	private enum StaticMapFailureTracker {
-		;
+	@SuppressWarnings({"StaticVariableUsedBeforeInitialization", "OverlyBroadThrowsClause"})
+	private static void restartTaskManager() throws Exception {
+		int tmi = lastTaskManagerIndexInMiniCluster.getAndIncrement();
+		miniCluster.terminateTaskExecutor(tmi).get();
+		miniCluster.startTaskExecutor();
+	}
 
-		private static final List<AtomicInteger> mapRestarts = new ArrayList<>(10);
-		private static final List<AtomicInteger> expectedMapRestarts = new ArrayList<>(10);
+	@FunctionalInterface
+	private interface FailureStrategy extends Serializable {
+		boolean failOrNot(int trackingIndex);
+	}
 
-		private static void reset() {
-			mapRestarts.clear();
-			expectedMapRestarts.clear();
-		}
+	private static class JoinedFailureStrategy implements FailureStrategy {
+		private static final long serialVersionUID = 1L;
 
-		private static int addNewMap() {
-			mapRestarts.add(new AtomicInteger(0));
-			expectedMapRestarts.add(new AtomicInteger(1));
-			return mapRestarts.size() - 1;
-		}
+		private final FailureStrategy[] failureStrategies;
 
-		private static void mapRestart(int index) {
-			mapRestarts.get(index).incrementAndGet();
+		private JoinedFailureStrategy(FailureStrategy ... failureStrategies) {
+			this.failureStrategies = failureStrategies;
 		}
 
-		private static void mapFailure(int index) {
-			expectedMapRestarts.get(index).incrementAndGet();
+		@Override
+		public boolean failOrNot(int trackingIndex) {
+			return Arrays
+				.stream(failureStrategies)
+				.anyMatch(failureStrategy -> failureStrategy.failOrNot(trackingIndex));
 		}
+	}
 
-		private static void verify() {
-			assertThat(collect(mapRestarts), is(collect(expectedMapRestarts)));
+	private static class RandomExceptionFailureStrategy extends AbstractRandomFailureStrategy {
+		private static final long serialVersionUID = 1L;
+
+		private RandomExceptionFailureStrategy(CoinToss coin) {
+			super(coin);
 		}
 
-		private static int[] collect(Collection<AtomicInteger> list) {
-			return list.stream().mapToInt(AtomicInteger::get).toArray();
+		@Override
+		void fail(int trackingIndex) {
+			StaticMapFailureTracker.mapFailure(trackingIndex);
+			throw new FlinkRuntimeException("BAGA-BOOM!!! The user function generated test failure.");
 		}
 	}
 
-	@FunctionalInterface
-	private interface FailureStrategy extends Serializable {
-		void failOrNot();
+	private static class RandomTaskExecutorFailureStrategy extends AbstractRandomFailureStrategy {
+		private static final long serialVersionUID = 1L;
+
+		private RandomTaskExecutorFailureStrategy(CoinToss coin) {
+			super(coin);
+		}
+
+		@Override
+		void fail(int trackingIndex) {
+			StaticMapFailureTracker.mapFailureWithBacktracking(trackingIndex);
+			try {
+				restartTaskManager();
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+			} catch (Exception e) {
+				ExceptionUtils.rethrow(e);
```

Comment:
If the shutdown of the TE fails we likely want to forward this exception directly to the test. As it stands we're just gonna restart the calling task, which may or may not result in a test failure (depending on whether the kill call went through). In any case it likely obfuscates things a bit.

Reply:
True, I will add tracking and verification of unexpected failures in `StaticMapFailureTracker`.